firewood/src/merkle.rs

```diff
@@ -51,27 +53,27 @@ fn nibbles_formatter<X: IntoIterator<Item = u8>>(nib: X) -> String {
 macro_rules! write_attributes {
     ($writer:ident, $node:expr, $value:expr) => {
         if !$node.partial_path.0.is_empty() {
-            write!(
-                $writer,
-                " pp={}",
-                nibbles_formatter($node.partial_path.0.clone())
-            )
-            .map_err(|e| FileIoError::from_generic_no_file(e, "write attributes"))?;
+            let pp_str = format!(" pp={}", nibbles_formatter($node.partial_path.0.clone()));
+            $writer
+                .write_all(pp_str.as_bytes())
+                .map_err(|e| FileIoError::from_generic_no_file(e, "write attributes"))?;
         }
         if !$value.is_empty() {
             match std::str::from_utf8($value) {
                 Ok(string) if string.chars().all(char::is_alphanumeric) => {
-                    write!($writer, " val={}", string)
+                    let val_str = format!(" val={}", string);
+                    $writer
+                        .write_all(val_str.as_bytes())
```

Comment:
this is unnecessary. you're constructing a heap string only to write it to the buffer when before it was correctly writing both directly to the writer.